File: core/src/qubitos/sme/measurement.py
# Copyright 2026 QubitOS Contributors
# SPDX-License-Identifier: Apache-2.0

"""Measurement-side helpers for the stochastic master equation solver.

Refs:
    - Wiseman and Milburn (2009), Quantum Measurement and Control, Ch. 4.
"""

from __future__ import annotations

import numpy as np
from numpy.typing import NDArray

from qubitos.lindblad import CollapseOperator

try:
    from agentbible import check_density_matrix as _ab_check_density_matrix
    from agentbible import check_finite as _ab_check_finite
    from agentbible import check_hermitian as _ab_check_hermitian

    _AGENTBIBLE_AVAILABLE = True
except ImportError:  # pragma: no cover - exercised by dependency-free installs
    _AGENTBIBLE_AVAILABLE = False

_DEBUG_VALIDATION = False


def set_debug_validation(enabled: bool) -> None:
    """Enable or disable thorough agentbible-based per-substep validation.

    When False (default), density-matrix surface checks use lightweight
    numpy finiteness + Hermiticity residue + trace deviation. When True,
    agentbible's check_density_matrix is used, which additionally validates
    physical-constant invariants at higher computational cost.

    The non-debug path is safe because the trajectory loop already
    symmetrizes and renormalizes every substep. Debug mode is useful for
    diagnosing integrator bugs during development.
    """
    global _DEBUG_VALIDATION
    _DEBUG_VALIDATION = enabled


def eigenvalue_bounds_2x2(rho: NDArray[np.complex128]) -> tuple[float, float]:
    """Closed-form 2x2 Hermitian eigenvalues (no eigendecomposition).

    For a 2x2 Hermitian density matrix written in Bloch form,
    eigenvalues are (trace/2) +/- sqrt((tr_diff/2)^2 + |b|^2).

    Returns (lambda_min, lambda_max). Matches the formula used by the
    Rust HAL backend in hal/src/sme/measurement.rs:eigenvalue_bounds_2x2.
    """
    a = rho[0, 0].real
    d = rho[1, 1].real
    b = complex(rho[0, 1])
    half_sum = 0.5 * (a + d)
    half_diff = 0.5 * (a - d)
    disc = np.sqrt(half_diff * half_diff + b.real * b.real + b.imag * b.imag)
    return float(half_sum - disc), float(half_sum + disc)


def nuclear_deviation_2x2(rho: NDArray[np.complex128]) -> float:
    """|nuclear norm - 1| for a 2x2 Hermitian density matrix.

    Nuclear norm = sum of absolute eigenvalues = |lambda_min| + |lambda_max|.
    Avoids the full SVD that numpy.linalg.norm(ord='nuc') would compute.
    """
    lam_min, lam_max = eigenvalue_bounds_2x2(symmetrize_density_matrix(rho))
    return abs(abs(lam_min) + abs(lam_max) - 1.0)


def effective_measurement_operator(
    collapse_ops: list[CollapseOperator],
    measurement_operator: NDArray[np.complex128] | None,
) -> NDArray[np.complex128]:
    """Resolve the effective measurement operator c used in H[c]ρ."""
    if measurement_operator is not None:
        return np.asarray(measurement_operator, dtype=np.complex128)
    if not collapse_ops:
        raise ValueError("A measurement_operator is required when collapse_ops is empty")
    primary = collapse_ops[0]
    return np.sqrt(primary.rate) * np.asarray(primary.matrix, dtype=np.complex128)


def measurement_expectation(
    measurement_operator: NDArray[np.complex128],
    rho: NDArray[np.complex128],
) -> float:
    """Return Tr[(c + c†)ρ] for the effective measurement operator c."""
    operator = measurement_operator + measurement_operator.conj().T
    return float(np.real(np.trace(operator @ rho)))


def measurement_superoperator(
    measurement_operator: NDArray[np.complex128],
    rho: NDArray[np.complex128],
) -> NDArray[np.complex128]:
    """Measurement innovation H[c]ρ = cρ + ρc† - Tr[(c + c†)ρ]ρ."""
    expectation = measurement_expectation(measurement_operator, rho)
    innovation = (
        measurement_operator @ rho + rho @ measurement_operator.conj().T - expectation * rho
    )
    validate_measurement_innovation(innovation)
    return innovation


def measurement_signal(
    measurement_operator: NDArray[np.complex128],
    rho: NDArray[np.complex128],
    eta: float,
    d_w: float,
    dt: float,
) -> float:
    """Homodyne photocurrent sample I(t) over one time step."""
    signal = np.sqrt(eta) * measurement_expectation(measurement_operator, rho) + d_w / dt
    check_finite_surface(np.array([signal]), name="measurement_signal")
    return float(signal)


def symmetrize_density_matrix(rho: NDArray[np.complex128]) -> NDArray[np.complex128]:
    """Project a matrix onto its Hermitian part."""
    return 0.5 * (rho + rho.conj().T)


def renormalize_density_matrix(rho: NDArray[np.complex128]) -> NDArray[np.complex128]:
    """Restore unit trace after a finite-timestep update."""
    trace = np.trace(rho)
    if abs(trace) < 1e-15:
        raise ValueError("Cannot renormalize density matrix with near-zero trace")
    return rho / trace


def project_positive_cone(rho: NDArray[np.complex128]) -> NDArray[np.complex128]:
    """Clamp 2x2 Bloch vector onto the unit ball and rebuild rho.

    Matches the Rust HAL backend logic in
    hal/src/sme/measurement.rs:project_positive_cone.
    Avoids an eigendecomposition for the common 2x2 case.
    """
    sym = symmetrize_density_matrix(rho)
    normalized = renormalize_density_matrix(sym)
    if normalized.shape[0] != 2:
        eigenvalues, eigenvectors = np.linalg.eigh(sym)
        clipped = np.clip(eigenvalues.real, 0.0, None)
        projected = eigenvectors @ np.diag(clipped.astype(np.complex128)) @ eigenvectors.conj().T
        return renormalize_density_matrix(symmetrize_density_matrix(projected))
    a = normalized[0, 0].real
    d = normalized[1, 1].real
    b = normalized[0, 1]
    x = 2.0 * b.real
    y = -2.0 * b.imag
    z = a - d
    norm = np.sqrt(x * x + y * y + z * z)
    if norm <= 1.0:
        return normalized
    x /= norm
    y /= norm
    z /= norm
    projected = np.zeros_like(normalized)
    projected[0, 0] = 0.5 * (1.0 + z)
    projected[1, 1] = 0.5 * (1.0 - z)
    projected[0, 1] = 0.5 * (x - 1.0j * y)
    projected[1, 0] = 0.5 * (x + 1.0j * y)
    return projected


def trace_deviation(rho: NDArray[np.complex128]) -> float:
    """Absolute deviation from unit trace."""
    return float(abs(np.trace(rho) - 1.0))


def trace_norm_deviation(rho: NDArray[np.complex128]) -> float:
    """Absolute deviation from unit trace norm.

    For 2x2 matrices uses the closed-form nuclear norm, ported from
    hal/src/sme/measurement.rs:trace_norm_deviation which uses
    eigenvalue_bounds_2x2. Falls back to full SVD for larger matrices.
    """
    if rho.shape[0] == 2:
        return nuclear_deviation_2x2(rho)
    return float(abs(np.linalg.norm(rho, ord="nuc") - 1.0))


def nonhermitian_residue(rho: NDArray[np.complex128]) -> float:
    """Maximum elementwise residue from Hermiticity."""
    return float(np.max(np.abs(rho - rho.conj().T)))


def has_positivity_violation(
    rho: NDArray[np.complex128],
    atol: float,
    min_eigenvalue: float | None = None,
) -> tuple[bool, float]:
    """Return whether rho has an eigenvalue smaller than -atol.

    If min_eigenvalue is provided, skips the eigendecomposition and uses it
    directly. This avoids redundant eigenvalue computation when the integrator
    already computed them for the retry decision.
    """
    if min_eigenvalue is None:
        if rho.shape[0] == 2:
            _, (lam_min, _) = True, eigenvalue_bounds_2x2(symmetrize_density_matrix(rho))
            min_eigenvalue = lam_min
        else:
            min_eigenvalue = float(
                np.min(np.linalg.eigvalsh(symmetrize_density_matrix(rho)).real)
            )
    return min_eigenvalue < -atol, min_eigenvalue


def validate_measurement_innovation(
    innovation: NDArray[np.complex128],
    atol: float = 1e-10,
) -> NDArray[np.complex128]:
    """Validate finiteness, Hermiticity, and trace-zero structure."""
    check_finite_surface(innovation, name="measurement_innovation")
    check_hermitian_surface(innovation, name="measurement_innovation", atol=atol)
    trace = complex(np.trace(innovation))
    if abs(trace) > atol:
        raise ValueError(f"measurement_innovation must be trace-zero, got {trace.real:.2e}")
    return innovation


def validate_trajectory_density_matrix(
    rho: NDArray[np.complex128],
    atol: float = 1e-6,
    min_eigenvalue: float | None = None,
) -> NDArray[np.complex128]:
    """Validate a stochastic per-trajectory density matrix surface.

    When min_eigenvalue is provided, reuses it for the positivity check
    rather than recomputing an eigendecomposition. The Eigenvalue is
    expected to come from the integrator step that produced rho (see
    euler_maruyama_step SMEStepResult.min_eigenvalue).

    Per-step validation uses lightweight checks by default. Use
    set_debug_validation(True) to enable agentbible for trajectory
    validation during development.
    """
    check_density_matrix_surface(
        rho, name="trajectory_density_matrix", atol=atol,
        min_eigenvalue=min_eigenvalue,
        use_agentbible=_DEBUG_VALIDATION,
    )
    return rho


def validate_ensemble_density_matrix(
    rho: NDArray[np.complex128],
    atol: float = 1e-10,
) -> NDArray[np.complex128]:
    """Validate an ensemble-averaged density matrix surface.

    Always uses the most thorough check available (including agentbible
    when installed) since this runs once at the end, not per-substep.
    """
    check_density_matrix_surface(
        rho, name="ensemble_density_matrix", atol=atol, use_agentbible=True,
    )
    return rho


def check_finite_surface(values: NDArray[np.complex128] | NDArray[np.float64], name: str) -> None:
    """Validate finiteness. Uses agentbible only when debug validation is on."""
    if _DEBUG_VALIDATION and _AGENTBIBLE_AVAILABLE:
        _ab_check_finite(np.asarray(values), name=name, strict=True)
        return
    if not np.all(np.isfinite(values)):
        raise ValueError(f"{name} contains non-finite values")


def check_hermitian_surface(
    matrix: NDArray[np.complex128],
    name: str,
    atol: float,
) -> None:
    """Validate Hermiticity. Uses agentbible only when debug validation is on."""
    if _DEBUG_VALIDATION and _AGENTBIBLE_AVAILABLE:
        _ab_check_hermitian(np.asarray(matrix), name=name, atol=atol, strict=True)
        return
    if nonhermitian_residue(matrix) > atol:
        raise ValueError(f"{name} is not Hermitian within atol={atol}")


def check_density_matrix_surface(
    matrix: NDArray[np.complex128],
    name: str,
    atol: float,
    min_eigenvalue: float | None = None,
    use_agentbible: bool = True,
) -> None:
    """Validate a density matrix surface.

    When use_agentbible is True and agentbible is available, delegates
    to agentbible for thorough invariant checking including physical-
    constant validation. When False, uses lightweight numpy checks
    (isfinite + Hermiticity residue + trace deviation + eigenvalue
    check with pre-computed eigenvalues when available).

    Per-substep trajectory validation passes use_agentbible=False for
    performance; ensemble validation (once at end) uses True.
    """
    if use_agentbible and _AGENTBIBLE_AVAILABLE:
        _ab_check_density_matrix(np.asarray(matrix), name=name, atol=atol, strict=True)
        return
    check_finite_surface(matrix, name=name)
    check_hermitian_surface(matrix, name=name, atol=atol)
    if trace_deviation(matrix) > atol:
        raise ValueError(f"{name} must have unit trace within atol={atol}")
    violation, min_eig = has_positivity_violation(
        matrix, atol, min_eigenvalue=min_eigenvalue,
    )
    if violation:
        raise ValueError(f"{name} has min eigenvalue {min_eig:.2e}")

File: hal/src/sme/measurement.rs
// Copyright 2026 QubitOS Contributors
// SPDX-License-Identifier: Apache-2.0

//! Measurement-side helpers for the stochastic master equation solver.

use agentbible::check::check_finite_array;
use ndarray::Array2;
use num_complex::Complex64;

use crate::lindblad::types::CollapseOperator;

/// Resolve the effective measurement operator c used in H[c]ρ.
pub fn effective_measurement_operator(
    collapse_ops: &[CollapseOperator],
    measurement_operator: Option<&Array2<Complex64>>,
) -> Result<Array2<Complex64>, String> {
    if let Some(op) = measurement_operator {
        return Ok(op.clone());
    }
    let primary = collapse_ops
        .first()
        .ok_or_else(|| "measurement_operator is required when collapse_ops is empty".to_string())?;
    Ok(primary.matrix.mapv(|z| z * primary.rate.sqrt()))
}

/// Tr[(c + c†)ρ] for the effective measurement operator c.
pub fn measurement_expectation(
    measurement_operator: &Array2<Complex64>,
    rho: &Array2<Complex64>,
) -> f64 {
    let operator = measurement_operator + &conjugate_transpose(measurement_operator);
    trace_real(&operator.dot(rho))
}

/// Measurement innovation H[c]ρ = cρ + ρc† - Tr[(c + c†)ρ]ρ.
pub fn measurement_superoperator(
    measurement_operator: &Array2<Complex64>,
    rho: &Array2<Complex64>,
) -> Result<Array2<Complex64>, String> {
    let expectation = measurement_expectation(measurement_operator, rho);
    let innovation = measurement_operator.dot(rho)
        + rho.dot(&conjugate_transpose(measurement_operator))
        - Complex64::new(expectation, 0.0) * rho;
    validate_measurement_innovation(&innovation, 1e-10)?;
    Ok(innovation)
}

/// One homodyne photocurrent sample for a timestep dt.
pub fn measurement_signal(
    measurement_operator: &Array2<Complex64>,
    rho: &Array2<Complex64>,
    eta: f64,
    d_w: f64,
    dt: f64,
) -> Result<f64, String> {
    let signal = eta.sqrt() * measurement_expectation(measurement_operator, rho) + d_w / dt;
    validate_finite_scalar(signal, "measurement_signal")?;
    Ok(signal)
}

/// Project a matrix onto its Hermitian part.
pub fn symmetrize_density_matrix(rho: &Array2<Complex64>) -> Array2<Complex64> {
    (rho + &conjugate_transpose(rho)) * Complex64::new(0.5, 0.0)
}

/// Restore unit trace after a finite-timestep update.
pub fn renormalize_density_matrix(rho: &Array2<Complex64>) -> Result<Array2<Complex64>, String> {
    let trace = trace_complex(rho);
    if trace.norm() < 1e-15 {
        return Err("Cannot renormalize density matrix with near-zero trace".into());
    }
    Ok(rho.mapv(|z| z / trace))
}

/// Clamp a 2×2 Bloch vector onto the unit ball and rebuild ρ.
pub fn project_positive_cone(rho: &Array2<Complex64>) -> Result<Array2<Complex64>, String> {
    let sym = symmetrize_density_matrix(rho);
    let normalized = renormalize_density_matrix(&sym)?;
    if normalized.nrows() != 2 || normalized.ncols() != 2 {
        return Err("positivity projection currently supports only 2x2 density matrices".into());
    }
    let a = normalized[[0, 0]].re;
    let d = normalized[[1, 1]].re;
    let b = normalized[[0, 1]];
    let mut x = 2.0 * b.re;
    let mut y = -2.0 * b.im;
    let mut z = a - d;
    let norm = (x * x + y * y + z * z).sqrt();
    if norm <= 1.0 {
        return Ok(normalized);
    }
    x /= norm;
    y /= norm;
    z /= norm;
    let mut projected = Array2::zeros((2, 2));
    projected[[0, 0]] = Complex64::new(0.5 * (1.0 + z), 0.0);
    projected[[1, 1]] = Complex64::new(0.5 * (1.0 - z), 0.0);
    projected[[0, 1]] = Complex64::new(0.5 * x, -0.5 * y);
    projected[[1, 0]] = Complex64::new(0.5 * x, 0.5 * y);
    Ok(projected)
}

/// Absolute deviation from unit trace.
pub fn trace_deviation(rho: &Array2<Complex64>) -> f64 {
    (trace_complex(rho) - Complex64::new(1.0, 0.0)).norm()
}

/// Absolute deviation from unit trace norm.
pub fn trace_norm_deviation(rho: &Array2<Complex64>) -> f64 {
    let sym = symmetrize_density_matrix(rho);
    let (lambda_min, lambda_max) = eigenvalue_bounds_2x2(&sym)
        .unwrap_or((trace_real(&sym) / 2.0, trace_real(&sym) / 2.0));
    (lambda_min.abs() + lambda_max.abs() - 1.0).abs()
}

/// Maximum elementwise residue from Hermiticity.
pub fn nonhermitian_residue(rho: &Array2<Complex64>) -> f64 {
    let diff = rho - &conjugate_transpose(rho);
    diff.iter().map(|z| z.norm()).fold(0.0, f64::max)
}

/// Return whether ρ has an eigenvalue smaller than -atol.
pub fn has_positivity_violation(rho: &Array2<Complex64>, atol: f64) -> (bool, f64) {
    let sym = symmetrize_density_matrix(rho);
    if let Some((lambda_min, _)) = eigenvalue_bounds_2x2(&sym) {
        return (lambda_min < -atol, lambda_min);
    }
    let min_diag = (0..sym.nrows())
        .map(|i| sym[[i, i]].re)
        .fold(f64::INFINITY, f64::min);
    (min_diag < -atol, min_diag)
}

/// Validate finiteness, Hermiticity, and trace-zero structure.
pub fn validate_measurement_innovation(
    innovation: &Array2<Complex64>,
    atol: f64,
) -> Result<(), String> {
    validate_finite_matrix(innovation, "measurement_innovation")?;
    validate_hermitian(innovation, "measurement_innovation", atol)?;
    let trace = trace_complex(innovation);
    if trace.norm() > atol {
        return Err(format!(
            "measurement_innovation must be trace-zero, got {:.2e} + {:.2e}i",
            trace.re, trace.im
        ));
    }
    Ok(())
}

/// Validate a stochastic per-trajectory density matrix surface.
pub fn validate_trajectory_density_matrix(rho: &Array2<Complex64>) -> Result<(), String> {
    validate_density_matrix_surface(rho, "trajectory_density_matrix", 1e-6)
}

/// Validate an ensemble-averaged density matrix surface.
pub fn validate_ensemble_density_matrix(rho: &Array2<Complex64>) -> Result<(), String> {
    validate_density_matrix_surface(rho, "ensemble_density_matrix", 1e-10)
}

fn validate_density_matrix_surface(
    rho: &Array2<Complex64>,
    name: &str,
    atol: f64,
) -> Result<(), String> {
    validate_finite_matrix(rho, name)?;
    validate_hermitian(rho, name, atol)?;
    if trace_deviation(rho) > atol {
        return Err(format!("{name} must have unit trace within atol={atol:.1e}"));
    }
    let (violation, min_eigenvalue) = has_positivity_violation(rho, atol);
    if violation {
        return Err(format!("{name} has min eigenvalue {min_eigenvalue:.2e}"));
    }
    Ok(())
}

fn validate_finite_matrix(matrix: &Array2<Complex64>, name: &str) -> Result<(), String> {
    let flat = flatten_real_imag(matrix);
    check_finite_array(&flat).map_err(|error| format!("{name}: {error}"))
}

fn validate_finite_scalar(value: f64, name: &str) -> Result<(), String> {
    if value.is_finite() {
        Ok(())
    } else {
        Err(format!("{name} is not finite"))
    }
}

fn validate_hermitian(matrix: &Array2<Complex64>, name: &str, atol: f64) -> Result<(), String> {
    if nonhermitian_residue(matrix) > atol {
        Err(format!("{name} is not Hermitian within atol={atol:.1e}"))
    } else {
        Ok(())
    }
}

fn conjugate_transpose(matrix: &Array2<Complex64>) -> Array2<Complex64> {
    matrix.t().mapv(|z| z.conj())
}

fn trace_complex(matrix: &Array2<Complex64>) -> Complex64 {
    (0..matrix.nrows()).fold(Complex64::new(0.0, 0.0), |acc, i| acc + matrix[[i, i]])
}

fn trace_real(matrix: &Array2<Complex64>) -> f64 {
    trace_complex(matrix).re
}

fn flatten_real_imag(matrix: &Array2<Complex64>) -> Vec<f64> {
    let mut flat = Vec::with_capacity(matrix.len() * 2);
    for value in matrix.iter() {
        flat.push(value.re);
        flat.push(value.im);
    }
    flat
}

fn eigenvalue_bounds_2x2(matrix: &Array2<Complex64>) -> Option<(f64, f64)> {
    if matrix.nrows() != 2 || matrix.ncols() != 2 {
        return None;
    }
    let a = matrix[[0, 0]].re;
    let d = matrix[[1, 1]].re;
    let b = matrix[[0, 1]];
    let half_sum = 0.5 * (a + d);
    let half_diff = 0.5 * (a - d);
    let disc = (half_diff * half_diff + b.norm_sqr()).sqrt();
    Some((half_sum - disc, half_sum + disc))
}

#[cfg(test)]
mod tests {
    use super::*;
    use approx::assert_relative_eq;

    fn plus_state() -> Array2<Complex64> {
        let half = Complex64::new(0.5, 0.0);
        let mut rho = Array2::zeros((2, 2));
        rho[[0, 0]] = half;
        rho[[0, 1]] = half;
        rho[[1, 0]] = half;
        rho[[1, 1]] = half;
        rho
    }

    #[test]
    fn measurement_superoperator_preserves_trace() {
        let op = CollapseOperator::amplitude_damping(50.0, "q0").unwrap();
        let measurement = effective_measurement_operator(&[op], None).unwrap();
        let innovation = measurement_superoperator(&measurement, &plus_state()).unwrap();
        assert_relative_eq!(trace_complex(&innovation).norm(), 0.0, epsilon = 1e-10);
    }

    #[test]
    fn project_positive_cone_repairs_negative_eigenvalue() {
        let mut bad = Array2::zeros((2, 2));
        bad[[0, 0]] = Complex64::new(1.05, 0.0);
        bad[[1, 1]] = Complex64::new(-0.05, 0.0);
        let projected = project_positive_cone(&bad).unwrap();
        let (violation, _) = has_positivity_violation(&projected, 1e-10);
        assert!(!violation);
        assert_relative_eq!(trace_real(&projected), 1.0, epsilon = 1e-10);
    }
}
